src/toxcore/onion/request_queue.rs

```diff
@@ -50,4 +50,28 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Returns the Instant of a PublicKey, used to check if the PK is timedout
+    pub fn get_ping_time<F>(&self, f: F) -> Option<&Instant>
+        where F: Fn(&T) -> bool {
+        let ping_times = self.ping_map.iter()
+            .filter_map(|(_ping_id, (ping_time, data))| {
+                if f(data) {
+                    Some(ping_time)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let len = ping_times.len();
+        if len == 0 {
+            None
+        } else if len > 1 {
+            warn!("There are multiple same pings to a peer");
+            None
```

Comment:
Do we need this warn? Why don't just use `find_map`? Also returning `None` here will lead to more duplications.